### PR TITLE
feat!: wrap the package generation in a Go cli

### PR DESCRIPTION
**Pull request by Skisocks**
# Changes
71 files changed so I'll explain what I've done. No changes have been made to the current setup, so won't have any effect on current pipelines. Will remove these once everything is migrated.

Squashed the commits as it was mostly me debugging stuff in the pipelines so added no value.
## CLI
* Wrapped package generation in a Go cli (has same functionality as previous setup)
  * Covers Python, Java, C#, & Angular
  * Designed with the intention of adding a specification generator command at a later time
* Added an integration test that creates & pushes mock packages for jx3-openapi-generation for testing (without killing the token)
* An example of how the new pipeline works can be seen [here](https://github.com/spring-financial-group/mqube-case-service/blob/test-opanpi/.lighthouse/jenkins-x/pullrequest.yaml#L19)
## Pipelines
* Installed the python dependencies to the Dockerfile
  * We can now move the python gen out of the horrible pipeline script. Want to test in a separate PR first.
* Removed `build-push-DANGER` and replaced with explicit steps that also push to latest/latest-preview.
  * Means pipeline definitions can use latest instead of having to bump the version each time
  * Also we can now remove `build-push-DANGER` as this is the only repo that uses it

### CODE REVIEW

File: .lighthouse/jenkins-x/integration.yaml [hunk 1]
```diff
@@ -0,0 +1,66 @@
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  creationTimestamp: null
+  name: peacock-release
```
Comment (tomhobson) [inline .lighthouse/jenkins-x/integration.yaml:5]: Don't think we wanna call it peacock release?

File: Dockerfile [hunk 2]
```diff
@@ -1,22 +1,35 @@
...
     && unzip -d /opt/gradle gradle-7.3.2-bin.zip \
     && rm gradle-7.3.2-bin.zip
 
 ENV PATH=$PATH:/opt/gradle/gradle-7.3.2/bin
 
+## Copy CLI binary & add to PATH
+COPY ./build/linux /jx3-openapi-generation
+ENV PATH "$PATH:/jx3-openapi-generation"
+
+# Add pipeline scripts & config files
```
Comment (tomhobson) [inline Dockerfile:32]: I see we're still building a bloody huge docker for this, we should look at cutting this down in future

File: pkg/packageGenerator/csharp/generator.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+	packageDir, err := g.FileIO.MkdirAll(filepath.Join(outputDir, g.GetPackageName()), 0755)
+	if err != nil {
+		return "", err
+	}
+
+	// Generate Package
+	err = g.Cmd.ExecuteAndLog(outputDir, "npx", "openapi-generator-cli", "generate",
+		"-i", specificationPath, "-g", "csharp-netcore", "-o", packageDir, "--git-user-id", g.RepoOwner,
+		"--git-repo-id", g.RepoName,
+		fmt.Sprintf("--additional-properties=targetFramework=netstandard2.1,packageName=%s,packageVersion=%s,netCoreProjectFile=true,optionalEmitDefaultValues=true,validatable=false", g.GetPackageName(), g.Version))
```
Comment (tomhobson) [inline pkg/packageGenerator/csharp/generator.go:47]: I think these things should be a config or something, they're pretty unreadable here?

File: pkg/packageGenerator/csharp/generator.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+func (g *Generator) GeneratePackage(specificationPath, outputDir string) (string, error) {
+	packageDir, err := g.FileIO.MkdirAll(filepath.Join(outputDir, g.GetPackageName()), 0755)
+	if err != nil {
+		return "", err
+	}
+
+	// Generate Package
+	err = g.Cmd.ExecuteAndLog(outputDir, "npx", "openapi-generator-cli", "generate",
```
Comment (tomhobson) [inline pkg/packageGenerator/csharp/generator.go:44]: Could definitely do this with a base config or something I think, sems like a lot of duplication?

File: pkg/packageGenerator/csharp/generator.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+	return packageDir, nil
+}
+
+func (g *Generator) GetPackageName() string {
+	return fmt.Sprintf("Mqube.%s.Client", g.ServiceName)
+}
+
+func (g *Generator) PushPackage(packageDir string) error {
+	solutionPath := fmt.Sprintf("./src/%s/bin/Release/**/*.nupkg", g.GetPackageName())
+	return g.Cmd.ExecuteAndLog(packageDir, "dotnet", "nuget", "push", solutionPath, "-s", "mqube.packages", "--skip-duplicate")
```
Comment (tomhobson) [inline pkg/packageGenerator/csharp/generator.go:71]: This is really the only custom bit I guess, could we use a base class for the other bits?

File: pkg/packageGenerator/generator_factory.go [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+	var generator domain.PackageGenerator
+	switch language {
+	case domain.CSharp:
+		generator = csharp.NewGenerator(f.Version, f.ServiceName, f.RepoOwner, f.RepoName)
+	case domain.Java:
+		generator = java.NewGenerator(f.Version, f.ServiceName, f.RepoOwner, f.RepoName)
+	case domain.Angular:
+		generator = angular.NewGenerator(f.Version, f.ServiceName, f.RepoOwner, f.RepoName)
+	case domain.Python:
+		generator = python.NewGenerator(f.Version, f.ServiceName, f.RepoOwner, f.RepoName, f.GitToken)
```
Comment (tomhobson) [inline pkg/packageGenerator/generator_factory.go:33]: feels like if we have a class we don't have to do a big case switch or whatever, we can just flick through the map until the language matches one we have or return an error?

**Review by tomhobson — CHANGES_REQUESTED** [review]
Think we can reduce the LOC a lot in this PR< mainly because the generators all seem to do the same thing, apart from the parameters and the push method?
WDYT?

File: Dockerfile (hunk 2 above)
Comment (Skisocks) [reply Dockerfile:32]: The alternative is we have an image per lang but then that would mean separating each lang into it's own pipeline step and I quite like the neatness of just calling the command once

File: pkg/packageGenerator/csharp/generator.go (hunk 3 above)
Comment (Skisocks) [reply pkg/packageGenerator/csharp/generator.go:47]: We can put all the lang configs into the openapitools.json. Haven't looked too deeply into it as I was just wanting to get it working as is.

File: pkg/packageGenerator/csharp/generator.go (hunk 5 above)
Comment (Skisocks) [reply pkg/packageGenerator/csharp/generator.go:71]: There are a number of files that also need to be brought in depending on the lang, I'll have a look and see what can be pulled out

File: pkg/packageGenerator/csharp/generator.go (hunk 3 above)
Comment (Skisocks) [reply pkg/packageGenerator/csharp/generator.go:47]: Apart from python, that uses a completely different package generator

**Review by tomhobson — APPROVED** [review]

**Pull request merged**